### PR TITLE
Add ENV variable handler to provide custom configuration to Cardano DB Sync

### PR DESCRIPTION
**Pull request by placek**
# Description

Because of [this issue](https://github.com/IntersectMBO/cardano-db-sync/issues/1777#issuecomment-2238847226) the GovTool team is forced to use some custom configuration file (Cardano DB Sync docker image is not provided with required flag enabled)

In order to do that I was in need to change the entrypoint of the DBSync docker image: The current one [hardcodes the configuration](https://github.com/IntersectMBO/cardano-db-sync/blob/master/nix/nixos/cardano-db-sync-service.nix#L221), so it is not adjustable.

This PR introduces a new environment variable, DB_SYNC_CONFIG, allowing users to specify a custom configuration file for the Cardano DB Sync service. The modification in `nix/nixos/cardano-db-sync-service.nix` adds a conditional check to use the custom configuration file if the environment variable is set, providing more flexibility in deployment scenarios and facilitating custom configurations without altering the base code. This change addresses a need for dynamic configuration management as specified in the recent user requirements.

# Checklist

- [x] Commit sequence broadly makes sense
- [x] Commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] Any changes are noted in the [changelog](https://github.com/IntersectMBO/cardano-db-sync/blob/master/db-sync/CHANGELOG.md)
- [ ] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) on version 0.10.1.0 (which can be run with `scripts/fourmolize.sh`)
- [x] Self-reviewed the diff

# Migrations

- [ ] The pr causes a [breaking change](https://github.com/IntersectMBO/cardano-db-sync/blob/master/doc/migrations.md) of type a,b or c
- [ ] If there is a breaking change, the pr includes a database migration and/or a fix process for old values, so that upgrade is possible
- [ ] Resyncing and running the migrations provided will result in the same database semantically

If there is a breaking change, especially a big one, please add a justification here. Please elaborate
more what the migration achieves, what it cannot achieve or why a migration is not possible.

### CODE REVIEW

**Comment by sgillespie** [comment]
This is a solid change, but I'd like to chat with you about this if possible. Are you by chance on the Intersect discord?

**Comment by placek** [comment]
Unfortunately I have no discord account.

I can freely discuss that here if it is not a problem for you.

**Comment by sgillespie** [comment]
> I can freely discuss that here if it is not a problem for you.

No problem. The current way of specifying a configuration is passing command-line args to the entrypoint (and not having `NETWORK` set). For example:

```
docker run -it --rm --volume "/home/sgillespie/dev/cardano-environments/sanchonet:/data" ghcr.io/intersectmbo/cardano-db-sync:sancho-5.1.0 --config /data/db-sync-config.json --socket-path /node-ipc/node.socket
```

So this is one approach, yours is another. Yet another approach would be to allow individual overrides via environment variables, eg `DB_SYNC_ENABLE_FUTURE_GENESIS=Y` or `DB_SYNC_INSERT_PRESET=only_gov`.

My question for you, as a user of cardano-db-sync-docker, is which of these approaches would you prefer?

**Comment by placek** [comment]
> The current way of specifying a configuration is passing command-line args to the entrypoint (and not having `NETWORK` set). For example:
> 
> ```
> docker run -it --rm --volume "/home/sgillespie/dev/cardano-environments/sanchonet:/data" \
>   ghcr.io/intersectmbo/cardano-db-sync:sancho-5.1.0 \
>   --config /data/db-sync-config.json \
>   --socket-path /node-ipc/node.socket
> ```

I tried this one before and had unexpected results, looking like the configuration I passed as an option was skipped. That's in fact why I created this PR.

> Yet another approach would be to allow individual overrides via environment variables, eg `DB_SYNC_ENABLE_FUTURE_GENESIS=Y` or `DB_SYNC_INSERT_PRESET=only_gov`.

I've checked this approach and frankly I don't know how to verify it's working:
```
$ docker exec -t my-cardano-db-sync-1 env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=488f2e073c59
TERM=xterm
DB_SYNC_ENABLE_FUTURE_GENESIS=Y
NETWORK=sanchonet
POSTGRES_HOST=postgres
POSTGRES_PORT=5432
RESTORE_RECREATE_DB=N
NIX_SSL_CERT_FILE=/nix/store/nnxyd3g8yxa2jcd5gy84crgcqav8pcya-nss-cacert-3.92/etc/ssl/certs/ca-bundle.crt
HOME=/root

$ docker exec -t my-cardano-db-sync-1 cat /nix/store/1ka1qz0xn71swc9zln7j6nm0hc0f9csa-db-sync-config.json
{"EnableLogMetrics":false,"EnableLogging":true,"NetworkName":"sanchonet","NodeConfigFile":"/nix/store/ybib87292glkclm4l5yqc0dq3fsy6w9l-config-sanchonet.json","PrometheusPort":8080,"RequiresNetworkMagic":"RequiresMagic","defaultBackends":["KatipBK"],"defaultScribes":[["StdoutSK","stdout"]],"minSeverity":"Info","options":{"cfokey":{"value":"Release-1.0.0"},"mapBackends":{},"mapSeverity":{"db-sync-node":"Info","db-sync-node.Mux":"Error","db-sync-node.Subscription":"Error"},"mapSubtrace":{"#ekgview":{"contents":[[{"contents":"cardano.epoch-validation.benchmark","tag":"Contains"},[{"contents":".monoclock.basic.","tag":"Contains"}]],[{"contents":"cardano.epoch-validation.benchmark","tag":"Contains"},[{"contents":"diff.RTS.cpuNs.timed.","tag":"Contains"}]],[{"contents":"#ekgview.#aggregation.cardano.epoch-validation.benchmark","tag":"StartsWith"},[{"contents":"diff.RTS.gcNum.timed.","tag":"Contains"}]]],"subtrace":"FilterTrace"},"#messagecounters.aggregation":{"subtrace":"NoTrace"},"#messagecounters.ekgview":{"subtrace":"NoTrace"},"#messagecounters.katip":{"subtrace":"NoTrace"},"#messagecounters.monitoring":{"subtrace":"NoTrace"},"#messagecounters.switchboard":{"subtrace":"NoTrace"},"benchmark":{"contents":["GhcRtsStats","MonotonicClock"],"subtrace":"ObservableTrace"},"cardano.epoch-validation.utxo-stats":{"subtrace":"NoTrace"}}},"rotation":{"rpKeepFilesNum":10,"rpLogLimitBytes":5000000,"rpMaxAgeHours":24},"setupBackends":["AggregationBK","KatipBK"],"setupScribes":[{"scFormat":"ScText","scKind":"StdoutSK","scName":"stdout","scRotation":null}]}

$ docker exec -t my-cardano-db-sync-1 cardano-db-sync version
cardano-db-sync 13.2.0.2 - linux-x86_64 - ghc-8.10
git revision 7dd3f60bf12c1d2853baf44e267a8dc3cb52bccf
```
In the logs I have no mention of this option either…

**Comment by placek** [comment]
Oh, I think I misuderstood you. You were suggesting a new approach to extend current functionality…

From my perspective it's better to provide a file based configuration that can be somehow shared with the Node. The ENV-based approach is also great, but since the DB Sync is heavily driven by it's dependencies (postgres & node) I think it's easier to manage this without ENVs.

**Comment by Cmdv** [comment]
> this file is generated by a Nix expression
🤦 
Well that blows my comments out of the window! Unfortunately due to having a hard time with nix playing well with MacOs. I run db-sync / cardano-node via a tmux script where I pass command line arguments to the location of the config files. So thought it would be perfect for docker to reflect that...

Oh my goodness just had a deeper look also didn't realise nix is used to generate the docker image/files too. Think @sgillespie you are best suited here 🏃 

sorry for the confusion!

**Comment by sgillespie** [comment]
> I tried this one before and had unexpected results, looking like the configuration I passed as an option was skipped. That's in fact why I created this PR.

Are you using the included docker-compose file? If so, I suspect it's setting `NETWORK` [even if it's not passed](https://github.com/IntersectMBO/cardano-db-sync/blob/master/docker-compose.yml#L59), which makes it launch the systemd service (rather than directly invoking it)

File: nix/nixos/cardano-db-sync-service.nix [hunk 1]
```diff
@@ -217,13 +217,23 @@ in {
...
             fi
 
-            exec ${exec} \
-                --config ${configFile} \
-                --socket-path "$CARDANO_NODE_SOCKET_PATH" \
-                --schema-dir ${self.schema} \
-                ${toString cfg.rtsArgs} \
-                ''${LEDGER_OPTS} \
-                ''${EXTRA_DB_SYNC_ARGS:-}
+            if [[ -z "''${CARDANO_DB_SYNC_CONFIG:-}" ]]; then
```
Comment (sgillespie) [inline nix/nixos/cardano-db-sync-service.nix:220]: Let's rename this to `DB_SYNC_CONFIG` to be consistent with the other environment variables

**Review by sgillespie — CHANGES_REQUESTED** [review]
After more thought, I think this way is better and less confusion. However, I'd also like the `docker-compose.yml` to be updated to include the new environment variable. Eg,

```
- CARDANO_DB_SYNC_CONFIG=${CARDANO_DB_SYNC_CONFIG:-}
```

I've also left a comment on the environment variable name itself

**Comment by placek** [comment]
> Are you using the included docker-compose file?

I am using the modified version of it.

> I suspect it's setting NETWORK [even if it's not passed](https://github.com/IntersectMBO/cardano-db-sync/blob/master/docker-compose.yml#L59), which makes it launch the systemd service (rather than directly invoking it)

So, you suggest that setting the `NETWORK` variable is the possible reason responsible for enforcing usage of some generated configuration (and not the one I directly state in docker command with `--config` option). Am I right?

I need to dive deeper to verify that. Anyways, there should be a note in documentation for that as well…

And this reminds me that I forgot to reflect my changes in documentation as well… let me fix that.

**Comment by sgillespie** [comment]
>   So, you suggest that setting the NETWORK variable is the possible reason responsible for enforcing usage of some generated configuration (and not the one I directly state in docker command with --config option). Am I right?

Yes, this is what I'm saying. 

> Anyways, there should be a note in documentation for that as well…

Agreed, we have a lot of work to do :)

> And this reminds me that I forgot to reflect my changes in documentation as well… let me fix that.

Don't forget to also update `docker-compose.yml`

**Comment by sgillespie** [comment]
@placek: Thanks for your work on this! We're going to go ahead and merge this in so we can get this in a hotfix. We'll handle the docs and docker-compose.yml file in subsequent PRs

**Review by sgillespie — APPROVED** [review]

**Pull request merged**